src/core/qgsvectorlayerexporter.cpp

```diff
@@ -120,6 +121,21 @@ QgsVectorLayerExporter::QgsVectorLayerExporter( const QString &uri,
     return;
   }
 
+  // If the result is a geopackage layer and there is already a field name FID requested which
+  // might contain duplicates, make sure to generate a new field with a unique name instead
+  // that will be filled by ogr with unique values.
+
+  // HACK sorry
+  if ( sinkOptions.testFlag( QgsFeatureSink::SinkFlag::RegeneratePrimaryKey ) && uri.endsWith( QLatin1String( ".gpkg" ) ) )
```

Comment:
I think you'll need to use decodeUri here -- couldn't a layername etc be sitting on the end of the uri?

Reply:
I can try. Those uris in all the variants are a miracle to me :/